File: lib/finance.ts
import Decimal from "decimal.js";

import type { TransactionType } from "@/db/schema";
import { addUtcMonths, monthsBetween, utcToDateInput } from "@/lib/dates";

export type FinancialEvent =
  | {
      kind: "transaction";
      date: string;
      createdAt: string;
      type: TransactionType;
      amountMinor: number;
    }
  | {
      kind: "valuation";
      date: string;
      createdAt: string;
      valueMinor: number;
    };

const POSITIVE_TYPES = new Set<TransactionType>([
  "opening_balance",
  "deposit",
  "interest",
  "dividend",
  "capital_gain",
  "purchase",
  "liability_increase",
]);

const NEGATIVE_TYPES = new Set<TransactionType>([
  "withdrawal",
  "capital_loss",
  "fee",
  "sale",
  "liability_payment",
]);

export function transactionEffect(
  type: TransactionType,
  amountMinor: number,
): bigint {
  const amount = BigInt(amountMinor);
  if (type === "manual_adjustment" || type === "transfer") return amount;
  if (POSITIVE_TYPES.has(type)) return amount < 0n ? -amount : amount;
  if (NEGATIVE_TYPES.has(type)) return amount > 0n ? -amount : amount;
  return 0n;
}

export function replayBalance(
  events: FinancialEvent[],
  throughDate?: string,
): bigint {
  const ordered = [...events]
    .filter((event) => !throughDate || event.date <= throughDate)
    .sort(
      (a, b) =>
        a.date.localeCompare(b.date) ||
        a.createdAt.localeCompare(b.createdAt) ||
        (a.kind === "transaction" ? -1 : 1),
    );

  let balance = 0n;
  for (const event of ordered) {
    if (event.kind === "valuation") {
      balance = BigInt(event.valueMinor);
    } else {
      balance += transactionEffect(event.type, event.amountMinor);
    }
  }
  return balance;
}

export type AccountFlowMetrics = {
  contributions: bigint;
  withdrawals: bigint;
  transfersIn: bigint;
  transfersOut: bigint;
  interest: bigint;
  dividends: bigint;
  fees: bigint;
  realizedGrowth: bigint;
};

export function calculateFlowMetrics(
  entries: Array<{ type: TransactionType; amountMinor: number | bigint }>,
): AccountFlowMetrics {
  const metrics: AccountFlowMetrics = {
    contributions: 0n,
    withdrawals: 0n,
    transfersIn: 0n,
    transfersOut: 0n,
    interest: 0n,
    dividends: 0n,
    fees: 0n,
    realizedGrowth: 0n,
  };

  for (const entry of entries) {
    const rawAmount = BigInt(entry.amountMinor);
    const amount = rawAmount < 0n ? -rawAmount : rawAmount;
    switch (entry.type) {
      case "opening_balance":
      case "deposit":
      case "purchase":
        metrics.contributions += amount;
        break;
      case "withdrawal":
      case "sale":
        metrics.withdrawals += amount;
        break;
      case "interest":
        metrics.interest += amount;
        break;
      case "dividend":
        metrics.dividends += amount;
        break;
      case "fee":
        metrics.fees += amount;
        break;
      case "capital_gain":
        metrics.realizedGrowth += amount;
        break;
      case "capital_loss":
        metrics.realizedGrowth -= amount;
        break;
      case "transfer":
        if (rawAmount >= 0n) metrics.transfersIn += amount;
        else metrics.transfersOut += amount;
        break;
      default:
        break;
    }
  }
  return metrics;
}

export function calculateNetWorthTotals(
  holdings: Array<{
    valueMinor: number | bigint;
    isLiability: boolean;
    included?: boolean;
  }>,
) {
  let assets = 0n;
  let liabilities = 0n;
  for (const holding of holdings) {
    if (holding.included === false) continue;
    const value = BigInt(holding.valueMinor);
    if (holding.isLiability) liabilities += value;
    else assets += value;
  }
  return { assets, liabilities, netWorth: assets - liabilities };
}

export function requiredMonthlyContribution(
  currentMinor: number | bigint,
  targetMinor: number | bigint,
  targetDate: Date,
  annualReturnBps = 0,
  fromDate = new Date(),
): bigint {
  const current = new Decimal(currentMinor.toString());
  const target = new Decimal(targetMinor.toString());
  if (current.greaterThanOrEqualTo(target)) return 0n;
  const months = Math.max(1, monthsBetween(fromDate, targetDate));
  const monthlyRate = new Decimal(annualReturnBps).div(10000).div(12);

  if (monthlyRate.isZero()) {
    return decimalToBigInt(
      target.minus(current).div(months),
      Decimal.ROUND_CEIL,
    );
  }

  const growthFactor = monthlyRate.plus(1).pow(months);
  const remainingAfterPrincipalGrowth = target.minus(current.mul(growthFactor));
  if (remainingAfterPrincipalGrowth.lessThanOrEqualTo(0)) return 0n;

  const contributionGrowthFactor = growthFactor.minus(1).div(monthlyRate);
  return decimalToBigInt(
    remainingAfterPrincipalGrowth.div(contributionGrowthFactor),
    Decimal.ROUND_CEIL,
  );
}

function decimalToBigInt(
  value: Decimal,
  rounding: Decimal.Rounding = Decimal.ROUND_HALF_UP,
) {
  return BigInt(value.toDecimalPlaces(0, rounding).toFixed(0));
}

export function monthlyPlanAmount(
  amountMinor: number | bigint,
  frequency: "weekly" | "monthly" | "quarterly" | "annually" | "custom",
): bigint {
  const amount = new Decimal(amountMinor.toString());
  switch (frequency) {
    case "weekly":
      return decimalToBigInt(amount.mul(52).div(12));
    case "quarterly":
      return decimalToBigInt(amount.div(3));
    case "annually":
      return decimalToBigInt(amount.div(12));
    case "monthly":
    case "custom":
      return BigInt(amountMinor);
  }
}

export function futureValueMinor(
  currentMinor: number | bigint,
  monthlyContributionMinor: number | bigint,
  annualReturnBps: number,
  months: number,
): bigint {
  if (months <= 0) return BigInt(currentMinor);
  const current = new Decimal(currentMinor.toString());
  const contribution = new Decimal(monthlyContributionMinor.toString());
  const monthlyRate = new Decimal(annualReturnBps).div(10000).div(12);

  if (monthlyRate.isZero()) {
    return decimalToBigInt(current.plus(contribution.mul(months)));
  }

  const growthFactor = monthlyRate.plus(1).pow(months);
  const principalGrowth = current.mul(growthFactor);
  const contributionGrowth = contribution
    .mul(growthFactor.minus(1))
    .div(monthlyRate);
  return decimalToBigInt(principalGrowth.plus(contributionGrowth));
}

export function futureValueWithContributionWindow(input: {
  currentMinor: number | bigint;
  monthlyContributionMinor: number | bigint;
  annualReturnBps: number;
  months: number;
  fromDate?: Date;
  contributionStart?: Date;
  contributionEnd?: Date | null;
}): bigint {
  let value = new Decimal(input.currentMinor.toString());
  const contribution = new Decimal(input.monthlyContributionMinor.toString());
  const monthlyRate = new Decimal(input.annualReturnBps).div(10000).div(12);
  const fromDate = input.fromDate ?? new Date();

  for (let month = 1; month <= input.months; month += 1) {
    value = value.mul(monthlyRate.plus(1));
    const contributionDate = addUtcMonths(fromDate, month);
    const contributionDay = utcToDateInput(contributionDate);
    const afterStart =
      !input.contributionStart ||
      contributionDay >= utcToDateInput(input.contributionStart);
    const beforeEnd =
      !input.contributionEnd ||
      contributionDay <= utcToDateInput(input.contributionEnd);
    if (afterStart && beforeEnd) value = value.plus(contribution);
  }
  return decimalToBigInt(value);
}

export function forecastCompletionWithContributionWindow(input: {
  currentMinor: number | bigint;
  targetMinor: number | bigint;
  monthlyContributionMinor: number | bigint;
  annualReturnBps: number;
  fromDate?: Date;
  contributionStart?: Date;
  contributionEnd?: Date | null;
}): Date | null {
  const fromDate = input.fromDate ?? new Date();
  if (BigInt(input.currentMinor) >= BigInt(input.targetMinor)) return fromDate;

  for (let month = 1; month <= 1200; month += 1) {
    if (
      futureValueWithContributionWindow({
        ...input,
        months: month,
        fromDate,
      }) >= BigInt(input.targetMinor)
    ) {
      return addUtcMonths(fromDate, month);
    }
  }
  return null;
}

export function forecastCompletionDate(
  currentMinor: number | bigint,
  targetMinor: number | bigint,
  monthlyContributionMinor: number | bigint,
  annualReturnBps: number,
  fromDate = new Date(),
): Date | null {
  if (BigInt(currentMinor) >= BigInt(targetMinor)) return fromDate;
  if (BigInt(monthlyContributionMinor) <= 0n && annualReturnBps <= 0)
    return null;

  for (let month = 1; month <= 1200; month += 1) {
    if (
      futureValueMinor(
        currentMinor,
        monthlyContributionMinor,
        annualReturnBps,
        month,
      ) >= BigInt(targetMinor)
    ) {
      return addUtcMonths(fromDate, month);
    }
  }
  return null;
}

export function projectGoalScenario(input: {
  currentMinor: number | bigint;
  targetMinor: number | bigint;
  monthlyContributionMinor: number | bigint;
  annualReturnBps: number;
  fromDate: Date;
  targetDate: Date;
}) {
  const monthsToTarget = Math.max(
    0,
    monthsBetween(input.fromDate, input.targetDate),
  );
  const current = BigInt(input.currentMinor);
  const monthlyContribution = BigInt(input.monthlyContributionMinor);
  const futureContributions = monthlyContribution * BigInt(monthsToTarget);
  const projectedAtTarget = futureValueMinor(
    current,
    monthlyContribution,
    input.annualReturnBps,
    monthsToTarget,
  );
  const forecastDate = forecastCompletionDate(
    current,
    input.targetMinor,
    monthlyContribution,
    input.annualReturnBps,
    input.fromDate,
  );

  return {
    monthsToTarget,
    projectedAtTarget,
    futureContributions,
    investmentGrowth: projectedAtTarget - current - futureContributions,
    forecastDate,
    reachesTarget: projectedAtTarget >= BigInt(input.targetMinor),
  };
}

export function goalTrackingStatus(input: {
  currentMinor: number | bigint;
  targetMinor: number | bigint;
  createdAt: Date;
  targetDate: Date;
  monthlyPlannedMinor: number | bigint;
  now?: Date;
}): "ahead" | "on_track" | "behind" {
  const now = input.now ?? new Date();
  if (BigInt(input.currentMinor) >= BigInt(input.targetMinor)) return "ahead";

  const totalMonths = Math.max(
    1,
    monthsBetween(input.createdAt, input.targetDate),
  );
  const elapsedMonths = Math.min(
    totalMonths,
    monthsBetween(input.createdAt, now),
  );
  const expected = new Decimal(input.targetMinor.toString())
    .mul(elapsedMonths)
    .div(totalMonths);
  const current = new Decimal(input.currentMinor.toString());
  const required = requiredMonthlyContribution(
    input.currentMinor,
    input.targetMinor,
    input.targetDate,
    0,
    now,
  );

  if (
    current.greaterThanOrEqualTo(expected.mul("1.05")) ||
    BigInt(input.monthlyPlannedMinor) > required
  ) {
    return "ahead";
  }
  if (
    current.lessThan(expected.mul("0.9")) ||
    BigInt(input.monthlyPlannedMinor) < required
  ) {
    return "behind";
  }
  return "on_track";
}

File: internal/domain/finance.go
package domain

import (
	"errors"
	"math/big"
	"sort"
)

var ErrEvent = errors.New("invalid financial event")

type TransactionType string

func TransactionEffect(kind TransactionType, amount Minor) (Minor, error) {
	value := big.NewInt(int64(amount))
	switch kind {
	case "opening_balance", "deposit", "interest", "dividend", "capital_gain", "purchase", "liability_increase":
		value.Abs(value)
	case "withdrawal", "capital_loss", "fee", "sale", "liability_payment":
		value.Abs(value).Neg(value)
	case "manual_adjustment", "transfer":
	default:
		return 0, ErrEvent
	}
	return checkedMinor(value)
}

type FinancialEvent struct {
	Kind        string          `json:"kind"`
	Date        string          `json:"date"`
	CreatedAt   string          `json:"createdAt"`
	Type        TransactionType `json:"type,omitempty"`
	AmountMinor Minor           `json:"amountMinor,omitempty"`
	ValueMinor  Minor           `json:"valueMinor,omitempty"`
}

// ReplayBalance copies input; valuations replace balances, never cash flow.
// Ties use transactions before valuations, preserving input order within a kind.
// Source dates/createdAt must be uniformly formatted UTC strings. Wide running
// sums preserve bigint semantics, with an int64 check at the returned boundary.
func ReplayBalance(events []FinancialEvent, throughDate string) (Minor, error) {
	ordered := make([]FinancialEvent, 0, len(events))
	for _, event := range events {
		if throughDate == "" || event.Date <= throughDate {
			ordered = append(ordered, event)
		}
	}
	sort.SliceStable(ordered, func(i, j int) bool {
		a, b := ordered[i], ordered[j]
		if a.Date != b.Date {
			return a.Date < b.Date
		}
		if a.CreatedAt != b.CreatedAt {
			return a.CreatedAt < b.CreatedAt
		}
		return a.Kind == "transaction" && b.Kind == "valuation"
	})
	balance := new(big.Int)
	for _, event := range ordered {
		switch event.Kind {
		case "valuation":
			balance.SetInt64(int64(event.ValueMinor))
		case "transaction":
			effect, err := TransactionEffect(event.Type, event.AmountMinor)
			if err != nil {
				return 0, err
			}
			balance.Add(balance, big.NewInt(int64(effect)))
		default:
			return 0, ErrEvent
		}
	}
	return checkedMinor(balance)
}

type FlowEntry struct {
	Type        TransactionType `json:"type"`
	AmountMinor Minor           `json:"amountMinor"`
}

type FlowMetrics struct {
	Contributions  Minor `json:"contributions"`
	Withdrawals    Minor `json:"withdrawals"`
	TransfersIn    Minor `json:"transfersIn"`
	TransfersOut   Minor `json:"transfersOut"`
	Interest       Minor `json:"interest"`
	Dividends      Minor `json:"dividends"`
	Fees           Minor `json:"fees"`
	RealizedGrowth Minor `json:"realizedGrowth"`
}

func CalculateFlowMetrics(entries []FlowEntry) (FlowMetrics, error) {
	sums := [8]big.Int{}
	for _, entry := range entries {
		if _, err := TransactionEffect(entry.Type, entry.AmountMinor); err != nil {
			return FlowMetrics{}, err
		}
		amount := new(big.Int).Abs(big.NewInt(int64(entry.AmountMinor)))
		index := -1
		switch entry.Type {
		case "opening_balance", "deposit", "purchase":
			index = 0
		case "withdrawal", "sale":
			index = 1
		case "transfer":
			index = 2
			if entry.AmountMinor < 0 {
				index = 3
			}
		case "interest":
			index = 4
		case "dividend":
			index = 5
		case "fee":
			index = 6
		case "capital_gain":
			index = 7
		case "capital_loss":
			index = 7
			amount.Neg(amount)
		}
		if index >= 0 {
			sums[index].Add(&sums[index], amount)
		}
	}
	result := FlowMetrics{}
	fields := []*Minor{&result.Contributions, &result.Withdrawals, &result.TransfersIn, &result.TransfersOut, &result.Interest, &result.Dividends, &result.Fees, &result.RealizedGrowth}
	for i, field := range fields {
		value, err := checkedMinor(&sums[i])
		if err != nil {
			return FlowMetrics{}, err
		}
		*field = value
	}
	return result, nil
}

type Holding struct {
	ValueMinor  Minor `json:"valueMinor"`
	IsLiability bool  `json:"isLiability"`
	Included    *bool `json:"included,omitempty"`
}

type NetWorthTotals struct {
	Assets      Minor `json:"assets"`
	Liabilities Minor `json:"liabilities"`
	NetWorth    Minor `json:"netWorth"`
}

func CalculateNetWorthTotals(holdings []Holding) (NetWorthTotals, error) {
	assets, liabilities := new(big.Int), new(big.Int)
	for _, holding := range holdings {
		if holding.Included != nil && !*holding.Included {
			continue
		}
		total := assets
		if holding.IsLiability {
			total = liabilities
		}
		total.Add(total, big.NewInt(int64(holding.ValueMinor)))
	}
	a, err := checkedMinor(assets)
	if err != nil {
		return NetWorthTotals{}, err
	}
	l, err := checkedMinor(liabilities)
	if err != nil {
		return NetWorthTotals{}, err
	}
	n, err := checkedMinor(new(big.Int).Sub(assets, liabilities))
	if err != nil {
		return NetWorthTotals{}, err
	}
	return NetWorthTotals{a, l, n}, nil
}
